File: kotlin/src/main/com/zlotindaniel/tddkatas/greeter/Greeter.kt
package com.zlotindaniel.tddkatas.greeter

import java.time.LocalTime

interface Logger {
	fun log(string: String)
}

class Greeter(private val logger: Logger? = null,
              private val getTime: () -> LocalTime) {

	fun greet(name: String): String {
		val txt = name.trim().capitalize()

		val result = when (getTime().hour) {
			in 6..11  -> "Good Morning $txt"
			in 12..17 -> "Hello $txt"
			in 18..21 -> "Good Evening $txt"
			else      -> "Good Night $txt"
		}

		logger?.log(result)
		return result
	}
}

File: go/pkg/greeter/greeter.go
package greeter

import (
	"fmt"
	"strings"
)

type GetHourOfDayFn func() int
type GreetingLogFn func(arg string)

type greeter struct {
	GreetingLogFn  GreetingLogFn
	getHourOfDayFn GetHourOfDayFn
}

func New(getHourOfDayFn GetHourOfDayFn) greeter {
	return greeter{getHourOfDayFn: getHourOfDayFn}
}

func (me *greeter) Greet(name string) string {
	title := strings.Title(strings.TrimSpace(name))
	hourOfDay := me.getHourOfDayFn()

	greeting := getGreeting(title, hourOfDay)

	if me.GreetingLogFn != nil {
		me.GreetingLogFn(greeting)
	}

	return greeting
}

func getGreeting(title string, hourOfDay int) string {
	if hourOfDay >= 6 && hourOfDay < 12 {
		return fmt.Sprintf("Good morning %s", title)
	} else if hourOfDay >= 18 && hourOfDay < 22 {
		return fmt.Sprintf("Good evening %s", title)
	} else if hourOfDay >= 22 || hourOfDay < 6 {
		return fmt.Sprintf("Good night %s", title)
	} else {
		return fmt.Sprintf("Hello %s", title)
	}
}
